// react-fresh/guard/decorator.spec.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var enzyme_1 = require("enzyme");
var component_1 = require("./component");
var decorator_1 = require("./decorator");
var util_1 = require("../util");
describe('PureGuard Decorator', function () {
    describe('UseGuard', function () {
        it('will put an instance of PureGuard with the guard over the component', function () {
            expect(util_1.shallowMockClassComponent.find(component_1.PureGuard).length).toEqual(0);
            var TestComponent = /** @class */ (function (_super) {
                __extends(TestComponent, _super);
                function TestComponent() {
                    return _super !== null && _super.apply(this, arguments) || this;
                }
                TestComponent = __decorate([
                    decorator_1.UseGuard('test')
                ], TestComponent);
                return TestComponent;
            }(util_1.MockClassComponent));
            expect(enzyme_1.shallow(react_1.default.createElement(TestComponent, null))
                .find(component_1.PureGuard)
                .at(0)
                .props()).toEqual({
                children: expect.any(Object),
                rule: 'test',
            });
        });
    });
    describe('GuardClass', function () {
        it('will throw an error for handle() by default', function () {
            var guardInstance = new decorator_1.GuardClass();
            expect(guardInstance.handle).toThrowError();
        });
    });
    describe('AddGuard', function () {
        it('will add a guard to the PureGuard guards', function () {
            expect(component_1.PureGuard.guards).toEqual({});
            var mockGuard = jest.fn().mockImplementation(function () { return false; });
            var TestGuard = /** @class */ (function () {
                function TestGuard() {
                    this.handle = mockGuard;
                }
                TestGuard = __decorate([
                    decorator_1.AddGuard('test')
                ], TestGuard);
                return TestGuard;
            }());
            component_1.PureGuard.guards.test();
            expect(mockGuard).toBeCalledTimes(1);
        });
    });
});
